src/runtime/model.cc

```diff
@@ -747,33 +732,114 @@ Tensor FFModel::create_constant(const int dims[],
       Predicate::TRUE_PRED, false, 0,
       FFConfig::get_hash_id(""));
   launcher.add_region_requirement(
-      RegionRequirement(tensor.part, 0/*projection id*/,
-                        WRITE_ONLY, EXCLUSIVE, tensor.region));
+      RegionRequirement(tensor->part, 0/*projection id*/,
+                        WRITE_ONLY, EXCLUSIVE, tensor->region));
   launcher.add_field(0, FID_DATA);
   FutureMap fm = runtime->execute_index_space(ctx, launcher);
   fm.wait_all_results();
   return tensor;
 }
 
+Tensor FFModel::create_tensor(
+    int numdim,
+    const int dims[],
+    DataType data_type,
+    const Op* op,
+    int idx,
+    bool create_grad)
+{
+  switch (numdim) {
+#define DIMFUNC(DIM) \
+    case DIM: \
+      return create_tensor<DIM>(dims, data_type, op, idx, create_grad);
+    LEGION_FOREACH_N(DIMFUNC)
+#undef DIMFUNC
+    default:
+      assert(false && "Unsupported dim!");
+  }
+}
+
+template<int NDIM>
+Tensor FFModel::create_tensor(
+    const int dims[],
+    DataType data_type,
+    const Op* owner_op,
+    int owner_idx,
+    bool create_grad)
+{
+  Tensor tensor = new TensorBase();
```

Comment:
Do users require to free the Tensor memory explicitly?

Reply:
Yes, we will decouple the creation of tensors and the mapping for them. For tensors that are parts of `FFModel`, FlexFlow will automatically creates partitioning and mapping in `compile`. Users need to explicitly call `ffmodel.map_tensor(tensor)` to map the tensors for full input dataset.

Reply:
Another significant change is that we now copy tensors by reference (instead of by value). This will allow us to performa graph transformations on FFModel.